[Wasm] JS modularization

pavelsavara opened the pull request:
### Modularized dotnet.js
- By using emcc options `MODULARIZE` and `EXPORT_ES6`, depending on `WasmEnableES6` to produce ES6 or CommonJS modules respectively.
- `WasmEnableES6` enables `WasmBuildNative` because `dotnet.js` need to be re-linked
- CommonJS version is able to be loaded into global namespace and behaves as before this change.
- Added new `es6/*.js` and `es6/*.js` files which are included in `dotnet.js` creation. Details are described in `modularize-dotnet.md`
- it exports `createDotnetRuntime` method instead of global `Module`
    - **for usage see `src\mono\sample\wasm\browser-es6\runtime.js`**
    - `createDotnetRuntime(factory: (exports) => DotnetModuleConfig)` 
    takes function which creates `Module`-like config.
    - `createDotnetRuntime` returns `Promise<{MONO, INTERNAL, BINDING, Module}>` exports.
- New `Module` extension `DotnetModuleConfig` has following new fields
    - `disableDotNet6Compatibility` - when `true` and if runtime script is loaded as module, it would not pollute global namespace.
    - `configSrc` - location of `mono-config.json` as defined by `MonoConfig` .ts type
    - `onConfigLoaded` - callback called early during runtime init, when `MonoConfig` was loaded but runtime didn't start yet.
    - `onDotnetReady` - callback called when runtime is ready and all artifacts were loaded from `MonoConfig`. Or not called at all when `onRuntimeInitialized` is overriden in Blazor.
- **Key aspects are documented in `src/mono/wasm/runtime/modularize-dotnet.md` of this PR.**

### Other changes
- Improved `dotnet.d.ts` it is now generated into version control too, so that we could observe how it evolves.
- Removed legacy `--testing` argument and simplified `is_testing` logic in the the samples.
- Added `browser-es6` sample app, which uses `WasmEnableES6=true` to compile `dotnet.js` as ES6 module.
- Added `browser-legacy` sample, which uses `dotnet.js` CommonJS module and loads it into global namespace.
- Added `package.json` into the nupkg
- Tested that Blazor works with CommonJS-global output
- Drafted how Blazor could use the new API in https://github.com/dotnet/aspnetcore/pull/38721/

### Out of scope
- switch to ES6 as default target 
- update of samples and unit tests to load ES6

### NodeJs Usage after the change
```
import { createRequire } from 'module';
import createDotnetRuntime from 'dotnet'
import { dirname } from 'path';
import { fileURLToPath } from 'url';
import { exit } from 'process';

const { BINDING } = await createDotnetRuntime(() => ({
    imports: {
        require: createRequire(import.meta.url)
    },
    scriptDirectory: dirname(fileURLToPath(import.meta.url)) + '/',
    disableDotnet6Compatibility: true,
    configSrc: "./mono-config.json"
}));

try {
    const app_args = process.argv.slice(2);
    const result = await BINDING.call_assembly_entry_point("Wasm.Console.TypeScript.Sample.dll", [app_args], "m");
    exit(result)
} catch (error) {
    console.log("WASM ERROR " + error);
}
```

am11 commented on src/mono/wasm/runtime-test.js:
```diff
@@ -102,51 +102,56 @@ if (typeof globalThis.performance === 'undefined') {
-        for (let variable in processedArguments.setenv) {
-            Module.config.environment_variables[variable] = processedArguments.setenv[variable];
-        }
+loadDotnet("./dotnet.js").then((createDotnetRuntime) => {
```
Do we need to capture the promise result `{ MONO, INTERNAL, BINDING, Module }` for line 152?

pavelsavara replied:
We pass it on line 134. 
BTW this file is still CommonJS loaded to global namespace but without global `Module`.
So it's bit ugly. Once I'm confident with ES6, I will switch this too and make it nicer.

pavelsavara replied:
The whole `loadDotnet` will go away and with top level await `{ MONO, INTERNAL, BINDING, Module }` would be top level too.

am11 replied:
Cool, and thanks for this work! I was trying it out on deno runtime with your branch: https://github.com/am11/runtime/commit/8dea0bb01cb5b137bbbfa643d894e0e055c40ea5, will follow-up once we are fully switched to modules. :)

The Venn's diagram of support matrices between https://caniuse.com/?search=modules and https://caniuse.com/?search=wasm seem to agree that it's safe to switch dotnet.js to module encapsulation.

kg commented on src/mono/sample/wasm/browser-es6/Wasm.Browser.Sample.csproj:
```diff
@@ -0,0 +1,31 @@
+  <PropertyGroup>
+    <_SampleProject>Wasm.Browser.Sample.csproj</_SampleProject>
+  </PropertyGroup>
+<!--
```
Please add comments explaining why this and the above property are commented out

pavelsavara replied:
Removed

kg commented on src/mono/sample/wasm/browser-es6/runtime.js:
```diff
@@ -0,0 +1,21 @@
+import createRuntime from './dotnet.js'
+
+export const { MONO, INTERNAL, BINDING, Module } = await createRuntime(({ MONO, INTERNAL, BINDING, Module }) => ({
```
This is a weird signature - do I understand it right and it's like this?
```createRuntime (callback : Function) : Promise<{MONO,INTERNAL,BINDING,Module}>```
and then callback is ```callback (namespaces : {MONO,INTERNAL,BINDING,Module}) : ConfigObject```? How are the namespaces visible to the config getter if the runtime hasn't loaded yet? What will happen if it calls methods on them?

kg commented on src/mono/sample/wasm/browser/runtime.js:
```diff
@@ -2,19 +2,19 @@
     onDotNetReady: function () {
         try {
-            App.init();
+            App.init({ MONO, INTERNAL, BINDING, Module });
```
this seems like a bad pattern to encourage, it means that if we added a 5th namespace it would not get passed to App.init. The argument containing the namespaces should just have a name and that object should be passed through to App.init as-is

pavelsavara replied:
yes, this will be better in ES6, where I would not have to do `.then()`.

kg commented on src/mono/wasm/build/README.md:
```diff
@@ -32,6 +32,8 @@ Implementation:
     - *after* any of the wasm build targets, use `AfterTargets="WasmBuildApp"` on that target
 - Avoid depending on this target, because it is available only when the workload is installed. Use `$(WasmNativeWorkload)` to check if it is installed.
 
+- `WasmEnableES6` will cause native re-link and produce `dotnet.js` as ES6 module.
```
Expand on this to specify that in es6 mode nothing will appear in the global namespace by default

kg replied:
We may also want to document somewhere (here?) that turning this on will break the debugger, and then we can remove that warning once we fix the debugger

kg commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -107,6 +107,9 @@
       <WasmBuildNative Condition="'$(RunAOTCompilation)' == 'true' and '$(RunAOTCompilationAfterBuild)' == 'true'">true</WasmBuildNative>
 
       <WasmBuildNative Condition="'$(WasmBuildNative)' == '' and @(NativeFileReference->Count()) > 0" >true</WasmBuildNative>
+      <!-- need to re-link dotnet.js when targeting legacy or CommonJS -->
```
Isn't this that you need to re-link if *not* targeting legacy/commonJS?

pavelsavara replied:
Fixed comment. This will be correct after we flip the default to ES6.

kg commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -223,8 +229,12 @@
       <_WasmRuntimePackSrcFile ObjectFile="$(_WasmIntermediateOutputPath)%(FileName).o" />
 
-      <_DotnetJSSrcFile Include="$(_WasmRuntimePackSrcDir)\*.js" Exclude="$(_WasmRuntimePackSrcDir)\*.iffe.js"/>
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.extpre.js" Kind="extern-pre-js"/>
```
I'm not sure if it belongs here, but a comment somewhere describing what each of these files (extpre.js, iffe.js, etc) actually does/is used for would be really valuable. I have no idea

pavelsavara replied:
Added `src\mono\wasm\runtime\modularize-dotnet.md`, more feedback welcome!

kg commented on src/mono/wasm/runtime-test.js:
```diff
@@ -102,51 +102,56 @@ if (typeof globalThis.performance === 'undefined') {
+loadDotnet("./dotnet.js").then((createDotnetRuntime) => {
+    return createDotnetRuntime(({ MONO, INTERNAL, BINDING, Module }) => ({
+        disableDotNet6Compatibility: true,
+        mainScriptUrlOrBlob: "dotnet.js",
```
should this be ```./dotnet.js``` for consistency with mono-config below?

pavelsavara replied:
this was some legacy thing deleted.

kg commented on src/mono/wasm/runtime-test.js:
```diff
@@ -333,24 +344,23 @@ async function loadDotnet(file) {
             document.head.appendChild(script);
-            return globalThis.Module;
+            while (true) {
+                if (globalThis.createDotnetRuntime) {
```
This infinite loop needs a bailout condition (a timeout is probably fine)

pavelsavara replied:
This whole `loadDotnet` loader will be replaced by single `import` in ES6 for all engines. Added timeout for now.

kg commented on src/mono/wasm/runtime-test.js:
```diff
@@ -333,24 +344,23 @@ async function loadDotnet(file) {
+                if (globalThis.createDotnetRuntime) {
+                    return globalThis.createDotnetRuntime;
+                }
+                await new Promise(resolve => setTimeout(resolve, 10));
```
Is this basically "loading started happening asynchronously at some point, and it will finish at some point but we're not sure when, so spin and check for it to finish every 10ms"? A comment explaining this loop would be good

kg commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -20,15 +20,16 @@ import {
     mono_wasm_pre_init,
-    mono_wasm_on_runtime_initialized
+    mono_wasm_on_runtime_initialized,
+    runtime_is_initialized
```
why no mono_wasm prefix?

pavelsavara replied:
Confusingly, there is also `mono_wasm_runtime_ready` and `mono_wasm_runtime_is_ready` for debugger, meaning something very different. 

Long term, I would like to drop `mono_wasm` prefix from internal code and leave it only on API.

Right now I added it for consistency sake.

kg commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -106,26 +107,50 @@ export const BINDING: BINDING = <any>{
 // it exports methods to global objects MONO, BINDING and Module in backward compatible way
 // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
-function export_to_emscripten(dotnet: any, mono: any, binding: any, internal: any, module: any): void {
+export const __exportAPI: any = (mono: any, binding: any, internal: any, module: any): void => {
```
this syntax is really weird, can we keep it as a function and then expose the function as __exportAPI separately

kg commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -288,4 +310,10 @@ export interface BINDING {
+    MONO: MONO,
+    BINDING: BINDING,
+    INTERNAL: any,
+    Module: any
```
Is there a specific reason why Module and INTERNAL don't have specific types here? I thought we had interfaces for them. I don't necessarily think they need it though

pavelsavara replied:
This will be visible in `dotnet.d.ts`, which is our published API. `INTERNAL` and `Module` are not part of that API contract.

kg commented on src/mono/wasm/runtime/modularize-dotnet.pre.js:
```diff
@@ -0,0 +1,11 @@
+const MONO = {}, BINDING = {}, INTERNAL = {};
+var ENVIRONMENT_IS_GLOBAL = typeof globalThis.Module === "object";
+if (ENVIRONMENT_IS_GLOBAL) {
+    globalThis.Module.ready = Module.ready;
```
Let's check to see whether a 'ready' already exists on Module here and if so... throw or something, since we've previously had people load us alongside another global-scoped emscripten module and then file confused issues about it.
Doing that is incorrect and they shouldn't expect it to work, we might as well fail early so they know what they did wrong.

kg replied:
We may actually want to check for a non-empty Module period and reject that, but I assume in some cases they're pre-creating it and putting stuff in there. We might want to check for some specific property emscripten puts there once it's loaded something

pavelsavara replied:
In both modularized and global `Module` it will be merged into implementation `Module`. Emscripten is mostly designed to let user shoot himself in to foot with any "override". I think we should stay consistent.

Rejecting non-empty Module/EmscriptenModuleConfig doesn't work, because that's the only startup API we and Emscripten give. 

Specifically `Module.ready` is always replaced by emscripten or us. I added the check.

kg commented on src/mono/wasm/wasm.proj:
```diff
@@ -71,7 +71,8 @@
       <_EmccCommonFlags Include="-s STRICT_JS=1" />      
-      <_EmccCommonFlags Include="-s MODULARIZE=1" Condition="'$(WasmEnableES6)' != 'false'" />
+      <_EmccCommonFlags Include="-s EXPORT_NAME=&quot;'createDotnetRuntime'&quot;" />
+      <_EmccCommonFlags Include="-s MODULARIZE=1"/>
```
It's intentional that this is no longer conditional, right? Because if ES6 isn't enabled we use a commonjs module instead?

pavelsavara replied:
Exactly.

kg reviewed (changes requested):
Looks great, thanks for doing all this work! You don't need to make all the changes I've requested here

radekdoulik commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -265,7 +275,16 @@
+    <ItemGroup Condition="'$(WasmBuildNative)' == 'true'">
+      <_MonoRuntimeComponentDontLink Include="libmono-component-diagnostics_tracing-static.a"/>
+    </ItemGroup>
+    <ItemGroup Condition="'$(WasmBuildNative)' == 'true' and '$(Configuration)' != 'Debug'">
```
We should handle the other case here as well, to avoid linking with both non-stub and stub components.

pavelsavara commented:
> This is a weird signature - do I understand it right and it's like this? `createRuntime (callback : Function) : Promise<{MONO,INTERNAL,BINDING,Module}>` and then callback is `callback (namespaces : {MONO,INTERNAL,BINDING,Module}) : ConfigObject`? How are the namespaces visible to the config getter if the runtime hasn't loaded yet? What will happen if it calls methods on them?

They are both receiving the same API object instances, i.e. same `MONO` instance.
In the callback `MONO` and others are empty objects and they become populated before the callbacks are called. 
The return from `createDotnetRuntime` is obviously assigned only after the promise is resolved.

The callback could also be written with `return {...}` syntax, but adding statements before that return is not useful exactly because `MONO` and such are just empty shells.

You can think about it as if `MONO` was populated very early, during exports.ts `__exportAPI` function. The actual path is bit more complex. 

The functionality in the `onConfigLoaded` is pretty limited as it's before full start of mono VM. It happens during emscripten `preInit` event, after fetched config.

The `onDotNetReady` is called after all assets are loaded , mapping to original `config.loaded_cb`. It happens during emscripten `onRuntimeInitialized` after monoVm init + globalization + assemblies. This also matches when the top level promise is resolved. The original emscripten `Module.ready` promise is replaced this this one.

pavelsavara commented:
Currently this [corrupts some of the xharness `testResults.xml` files](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-5baefdff0f8047a3be/Microsoft.Extensions.Configuration.Json.Tests/1/console.c7973149.log?sv=2019-07-07&se=2021-11-28T21%3A18%3A56Z&sr=c&sp=rl&sig=sj1w45JheuB%2FHMJet2J26%2B1%2BpKiBVyYhyTnLUDNzLRs%3D). It seems to me it could be timing or something with `console`.

pavelsavara commented:
I added `src/mono/wasm/runtime/modularize-dotnet.md` which is probably worth reading for this PR.

pavelsavara commented:
test failure `runtime.js 2:34 Uncaught SyntaxError: Unexpected reserved word` [helix](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-3820174da89c49819f/Wasm.Browser.Sample/1/console.7c7ebc72.log?sv=2019-07-07&se=2021-11-29T22%3A24%3A35Z&sr=c&sp=rl&sig=cx3ByezgWVD%2B2PuK9JU%2BqE%2BHQeaOrP8N5hKf0Uk51UY%3D)

kg commented on src/mono/wasm/runtime/modularize-dotnet.extpostcjs.js:
```diff
@@ -0,0 +1,4 @@
+
+if (typeof globalThis.Module === "object") {
```
What's this doing?

kg reviewed (approved):
Looks good to me, assuming no one else finds any problems!

pavelsavara commented:
@radical please review the msbuild parts.

maraf reviewed (approved):
I only have a slight feeling that from user point of view, this syntax may seem a bit odd.

```
export const { MONO, BINDING } = await createDotnetRuntime(({ MONO, BINDING, Module }) => ...);
```

I mean "what is the difference between left MONO and right MONO?"

Anyway, I can't find a better solution.

radical reviewed (dismissed):
Reviewed the parts that I could

radical commented:
> Currently this [corrupts some of the xharness `testResults.xml` files](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-5baefdff0f8047a3be/Microsoft.Extensions.Configuration.Json.Tests/1/console.c7973149.log?sv=2019-07-07&se=2021-11-28T21%3A18%3A56Z&sr=c&sp=rl&sig=sj1w45JheuB%2FHMJet2J26%2B1%2BpKiBVyYhyTnLUDNzLRs%3D). It seems to me it could be timing or something with `console`.

The `testResults.xml` file is incomplete in those instances. `WASM EXIT` is the last line written, which indicates that the tests have run, and the xml file has been written too. I'm guessing that this is getting written too soon.

radical commented:
The AOT `System.Runtime.Tests` fail to run with ([log](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-d1168ded3ead41889f/System.Runtime.Tests/1/console.a502995c.log?sv=2019-07-07&)):

```
[22:32:58] info: Running v8 --expose_wasm --stack-trace-limit=1000 runtime.js -- --run WasmTestRunner.dll System.Runtime.Tests.dll -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing
                 
[22:32:58] fail: console.error: Error executing file
                 
[22:32:58] fail: console.error: failed to load the dotnet.js file
                 
[22:32:58] info: ./dotnet.js:3: SyntaxError: Cannot use 'import.meta' outside a module
[22:32:58] info:   var _scriptDir = import.meta.url;
[22:32:58] info:                           ^^^^
[22:32:58] info: SyntaxError: Cannot use 'import.meta' outside a module
[22:32:58] info:     at loadScript (runtime.js:361:24)
[22:32:58] info:     at loadDotnet (runtime.js:369:12)
[22:32:58] info:     at runtime.js:105:1
[22:32:58] info: 
[22:32:58] info: Incoming arguments: --run WasmTestRunner.dll System.Runtime.Tests.dll -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing
[22:32:58] info: Application arguments: --run WasmTestRunner.dll System.Runtime.Tests.dll -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing
```

Also, the app then incorrectly exits with exit code 0, thus it records as tests having passed. This should be fixed so if such an error occurs, then it returns a non-zero exit code.

radical commented:
Also, make sure to run *all* the wasm tests. A lot of them are disabled now for PRs, and run fully only for rolling builds, generally conditioned with `isFullMatrix` in `eng/pipelines/runtime.yml`, and `eng/pipelines/runtime-staging.yml`.

And cc @thaystg to confirm that debugging still works, and the tests pass too.

kg commented:
> 
> 
> Also, make sure to run _all_ the wasm tests. A lot of them are disabled now for PRs, and run fully only for rolling builds, generally conditioned with `isFullMatrix` in `eng/pipelines/runtime.yml`, and `eng/pipelines/runtime-staging.yml`.
> 
> And cc @thaystg to confirm that debugging still works, and the tests pass too.

Is there a way for us to use a command to manually run the full set?

radical commented:
> Is there a way for us to use a command to manually run the full set?

Not yet. AFAIU, @akoeplinger is working on adding support for manually triggering these jobs on PRs.

radical commented:
Something like [this](https://gist.github.com/radical/6949d829efe4d8372467963968f994f7) should do it as a temporary patch for the PR.

pavelsavara commented:
> [22:32:58] info: ./dotnet.js:3: SyntaxError: Cannot use 'import.meta' outside a module
Fixed in `WasmApp.Native.targets`

> Also, the app then incorrectly exits with exit code 0
When dotnet.js is not loaded, we don't have shell `exit()` yet. I fixed in on xharness+browser combination, I think.

radical commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -222,8 +228,26 @@
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.postes6.js" Kind="post-js" Condition="'$(WasmEnableES6)' == 'true'"/>
+
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.extpost.js" Kind="extern-post-js"/>
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.extpostcjs.js" Kind="extern-post-js" Condition="'$(WasmEnableES6)' != 'true'"/>
```
can we move all these into two separate folders, `cjs`, and `es6`?

radical commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -267,7 +291,20 @@
 
-  <Target Name="_WasmSelectRuntimeComponentsForLinking" Condition="'$(WasmNativeWorkload)' == true" DependsOnTargets="_MonoSelectRuntimeComponents" />
+  <Target Name="_WasmSelectRuntimeComponentsForLinkingWorkload" Condition="'$(WasmNativeWorkload)' == 'true'" DependsOnTargets="_MonoSelectRuntimeComponents" />
+  <Target Name="_WasmSelectRuntimeComponentsForLinkingInTree" Condition="'$(WasmNativeWorkload)' != 'true'" >
```
this belongs in the in-tree targets file

kg commented on src/mono/wasm/runtime-test.js:
```diff
@@ -248,7 +248,9 @@ function set_exit_code(exit_code, reason) {
     if (is_browser) {
         // tell xharness stream procesor we are done
-        originalConsole.log("WASM EXIT " + exit_code);
+        setTimeout(() => {
```
Does this need to be 200 specifically? Or are you still troubleshooting

pavelsavara replied:
I'm still troubleshooting.
I still don't know how the C# code could stop in the middle of reporting TestResults.xml
It was missing even [`After RunTests`](https://github.com/dotnet/runtime/pull/61313/files#diff-86185eb9de15ff10811d033f377f65f9ca6ee32bc5d90d95b7f95b66fd2aae46R75) log line on previous build

radical commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -267,8 +296,9 @@
-  <Target Name="_WasmSelectRuntimeComponentsForLinking" Condition="'$(WasmNativeWorkload)' == true" DependsOnTargets="_MonoSelectRuntimeComponents" />
-
+  <Target Name="_WasmSelectRuntimeComponentsForLinking" Condition="'$(WasmNativeWorkload)' == 'true'" DependsOnTargets="_MonoSelectRuntimeComponents" />
+  <Target Name="_WasmSelectRuntimeComponentsForLinking" Condition="'$(WasmNativeWorkload)' != 'true'" DependsOnTargets="_WasmSelectRuntimeComponentsForLinkingInTree" />
```
We should include the mono targets `src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk` with the tests, like we do for other tasks, and targets. So, this would work automatically, and we wouldn't need to patch here.

pavelsavara replied:
@radical could you please point out all the places ? I'm quite confused about it. There are probably few projects created from template during the unit test and I don't know if it belongs there or not. Or I don't know that they exist.

radical replied:
https://github.com/dotnet/runtime/blob/main/src/mono/wasm/build/README.md#updating-dependencies-needed-for-building-wasm-apps

pavelsavara replied:
We agreed with @radical to put back the temporary hack and keep https://github.com/dotnet/runtime/issues/61294 open for him to fix it properly.

radical commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -222,8 +228,31 @@
 
-      <_DotnetJSSrcFile Include="$(_WasmRuntimePackSrcDir)\*.js" Exclude="$(_WasmRuntimePackSrcDir)\*.iffe.js"/>
-      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.iffe.js" Kind="pre-js" />
+      <!-- See src\mono\wasm\runtime\modularize-dotnet.md -->
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.extpre.js"      Kind="extern-pre-js"/>
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\cjs\*.extpre.js"  Kind="extern-pre-js"  Condition="'$(WasmEnableES6)' != 'true'"/>
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\es6\*.extpre.js"  Kind="extern-pre-js"  Condition="'$(WasmEnableES6)' == 'true'"/>
+
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.pre.js"         Kind="pre-js" />
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\cjs\*.pre.js"     Kind="pre-js"         Condition="'$(WasmEnableES6)' != 'true'"/>
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\es6\*.pre.js"     Kind="pre-js"         Condition="'$(WasmEnableES6)' == 'true'"/>
+
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.iffe.js"        Kind="pre-js" />
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\cjs\*.iffe.js"    Kind="pre-js"         Condition="'$(WasmEnableES6)' != 'true'"/>
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\es6\*.iffe.js"    Kind="pre-js"         Condition="'$(WasmEnableES6)' == 'true'"/>
+
+      <_DotnetJSSrcFile Include="$(_WasmRuntimePackSrcDir)\*.lib.js"         Kind="js-library"/>
+      <_DotnetJSSrcFile Include="$(_WasmRuntimePackSrcDir)\cjs\*.lib.js"     Kind="js-library"     Condition="'$(WasmEnableES6)' != 'true'"/>
+      <_DotnetJSSrcFile Include="$(_WasmRuntimePackSrcDir)\es6\*.lib.js"     Kind="js-library"     Condition="'$(WasmEnableES6)' == 'true'"/>
+
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.post.js"        Kind="post-js" />
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\cjs\*.post.js"    Kind="post-js"        Condition="'$(WasmEnableES6)' != 'true'"/>
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\es6\*.post.js"    Kind="post-js"        Condition="'$(WasmEnableES6)' == 'true'"/>
+
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.extpost.js"     Kind="extern-post-js"/>
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\cjs\*.extpost.js" Kind="extern-post-js" Condition="'$(WasmEnableES6)' != 'true'"/>
+      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\es6\*.extpost.js" Kind="extern-post-js" Condition="'$(WasmEnableES6)' == 'true'"/>
+
```
Maybe something like:
```xml
      <JSFileType Include="extpre.js"  Kind="extern-pre-js" />
      <JSFileType Include="pre.js"     Kind="pre-js" />
      <JSFileType Include="iffe.js"    Kind="pre-js" />
      <JSFileType Include="lib.js"     Kind="js-library" />
      <JSFileType Include="post.js"    Kind="post-js" />
      <JSFileType Include="extpost.js" Kind="extern-post-js" />

      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\*.%(JSFileType.Identity)"     Kind="%(JSFileType.Kind)" />
      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\cjs\*.%(JSFileType.Identity)" Kind="%(JSFileType.Kind)" Condition="'$(WasmEnableES6)' != 'true'" />
      <_WasmExtraJSFile Include="$(_WasmRuntimePackSrcDir)\es6\*.%(JSFileType.Identity)" Kind="%(JSFileType.Kind)" Condition="'$(WasmEnableES6)' == 'true'" />
```

radical commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -351,10 +381,12 @@
       <_EmccLinkStepArgs Include="--js-library &quot;%(_DotnetJSSrcFile.Identity)&quot;" />
       <_WasmLinkDependencies Include="@(_DotnetJSSrcFile)" />
 
-      <_EmccLinkStepArgs Include="--js-library &quot;%(_WasmExtraJSFile.Identity)&quot;" Condition="'%(_WasmExtraJSFile.Kind)' == 'js-library'" />
-      <_EmccLinkStepArgs Include="--pre-js &quot;%(_WasmExtraJSFile.Identity)&quot;"     Condition="'%(_WasmExtraJSFile.Kind)' == 'pre-js'" />
-      <_EmccLinkStepArgs Include="--post-js &quot;%(_WasmExtraJSFile.Identity)&quot;"    Condition="'%(_WasmExtraJSFile.Kind)' == 'post-js'" />
-      <_WasmLinkDependencies Include="@(_WasmExtraJSFile)" Condition="'%(_WasmExtraJSFile.Kind)' == 'js-library' or '%(_WasmExtraJSFile.Kind)' == 'pre-js' or '%(_WasmExtraJSFile.Kind)' == 'post-js'" />
+      <_EmccLinkStepArgs Include="--extern-pre-js &quot;%(_WasmExtraJSFile.Identity)&quot;"  Condition="'%(_WasmExtraJSFile.Kind)' == 'extern-pre-js'" />
+      <_EmccLinkStepArgs Include="--pre-js &quot;%(_WasmExtraJSFile.Identity)&quot;"         Condition="'%(_WasmExtraJSFile.Kind)' == 'pre-js'" />
+      <_EmccLinkStepArgs Include="--js-library &quot;%(_WasmExtraJSFile.Identity)&quot;"     Condition="'%(_WasmExtraJSFile.Kind)' == 'js-library'" />
+      <_EmccLinkStepArgs Include="--post-js &quot;%(_WasmExtraJSFile.Identity)&quot;"        Condition="'%(_WasmExtraJSFile.Kind)' == 'post-js'" />
+      <_EmccLinkStepArgs Include="--extern-post-js &quot;%(_WasmExtraJSFile.Identity)&quot;" Condition="'%(_WasmExtraJSFile.Kind)' == 'extern-post-js'" />
+      <_WasmLinkDependencies Include="@(_WasmExtraJSFile)" Condition="'%(_WasmExtraJSFile.Kind)' == 'js-library' or '%(_WasmExtraJSFile.Kind)' == 'pre-js' or '%(_WasmExtraJSFile.Kind)' == 'post-js' or '%(_WasmExtraJSFile.Kind)' == 'extern-pre-js' or '%(_WasmExtraJSFile.Kind)' == 'extern-post-js'" />
```
and
```xml
      <_EmccLinkStepArgs Include="--%(_WasmExtraJSFile.Kind) &quot;%(_WasmExtraJSFile.Identity)&quot;" Condition="'%(_WasmExtraJSFile.Kind)' != ''" />
      <_WasmLinkDependencies Include="@(_WasmExtraJSFile)" />
```

Elringus commented:
@pavelsavara Thank you for the work on modularization, it's a very useful feature! 

Can you please clarify, once this is ready, will we be able to `dotnet publish -r browser` (or something similar) and get `dotnet.wasm` and `dotnet.js` ready to be imported as modules without requirements for DOM or any other environment-specific APIs (fetch, path, etc)?

My current issue with `dotnet publish -r browser-wasm` (or Microsoft.NET.Sdk.BlazorWebAssembly in general) is that it outputs wasn and js that can only be used in browser, while I'm looking to embed them with a UMD js library to be used in a custom environment, where neither dom nor nodejs APIs are available.

pavelsavara commented:
> Can you please clarify, once this is ready, will we be able to `dotnet publish -r wasm` (or something similar) and get `dotnet.wasm` and `dotnet.js` ready to be imported as modules without requirements for DOM or any other environment-specific APIs (fetch, path, etc)?

That is my vision, but I'm not sure we would get there 100% in this PR.

The emscripten itself contains code which `require('path')`, `fs`, `crypto`, `ws`, `url` some of it conditionally on the detected platform. For ES6 I will replace `require` with `import` so that it would work in NodeJS as ES6 module.

I'm not aware of DOM requirement in the current `dotnet.js`, we are already running it in V8 shell without DOM. Could you please be more specific ?

`fetch` equivalent is required so that dotnet could load assemblies and other files. We will provide extension point, so that you could provide your own implementation.

Elringus commented:
> I'm not aware of DOM requirement in the current `dotnet.js`, we are already running it in V8 shell without DOM. Could you please be more specific ?

I remember webpack arguing about "window" and "document" when attempting to pack it as UMD module. Those are present in dotnet.js, but are probably not used in non-browser environments.

I was thinking about compiling emscripten with "shell" [environment](https://github.com/emscripten-core/emscripten/blob/main/src/settings.js#L641) and `--closure` flag, so that the generated .js is stripped out of any unused code, but the ems version dotnet uses (2.0.23) has a bug with closure, unfortunately.

I'm currently compiling for web, as it produces minimum extra code and post-processing the generated dotnet.js further removing the unused statements: https://github.com/Elringus/DotNetJS/blob/126623fdfc1608466df1a8ec33d5089b56fa4209/Runtime/scripts/compile-runtime.sh This way I can pack dotnet.js with a UMD library and everything seems to work fine both in browser and node. The wasm and dll binaries are provided by the consumer of the library.

pavelsavara committed:
- modularized `dotnet.js`
    - By using emcc options `MODULARIZE` and `EXPORT_ES6`, depending on `WasmEnableES6` to produce ES6 or CommonJS modules respectively.
    - `WasmEnableES6` enables `WasmBuildNative` because `dotnet.js` need to be re-linked
    - CommonJS version is able to be loaded into global namespace and behaves as before this change.
    - Added new `modularize-dotnet.*.js` files which are included in `dotnet.js` creation.
    - it exports `createDotnetRuntime` method instead of global `Module`
        - `createDotnetRuntime` takes function which creates `Module`-like config `(exports) => EmscriptenModuleConfig` .
        - `createDotnetRuntime` returns `Promise<{MONO, INTERNAL, BINDING, Module}>` exports.
- Added `browser-es6` sample app, which uses `WasmEnableES6=true` to compile `dotnet.js` as ES6 module.
- Added `browser-legacy` sample, which uses `dotnet.js` CommonJS module and loads it into global namespace.
- Attempt to fix In-Tree `_WasmSelectRuntimeComponentsForLinking` when `WasmBuildNative`
- feedback from Radek & Kate & Ankit
- rebased

pavelsavara commented:
Last build failed with `http://127.0.0.1:44591/main.js 2:34 Uncaught SyntaxError: Unexpected reserved word` [log](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-986c20762e6d411c97/Wasm.Browser.Sample/1/console.68204d57.log?sv=2019-07-07&se=2021-12-07T08%3A19%3A43Z&sr=c&sp=rl&sig=T9KhiMJcvMNdk%2FbLLvs4EISY%2FdoVmuJXKLNfsB%2FOuNE%3D)

And `The target "_WasmSelectRuntimeComponentsForLinkingInTree" does not exist in the project. [/datadisks/disk1/work/B4CF09D1/w/B747099E/e/publish/ProxyProjectForAOTOnHelix.proj]` [log](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-2c07ef095af04ca18a/Invariant.Tests/1/console.e0611df9.log?sv=2019-07-07&se=2021-12-07T08%3A37%3A24Z&sr=c&sp=rl&sig=G3%2FFV3h9BedXm2bU3PDiFns%2BKJUdpuY8WcmbMlZReJg%3D)

kg commented:
Doubt it's the problem but I think the bindings JS generators don't check for reserved words, I should make a todo note to fix that

pavelsavara commented:
It's most likely that there is some mistake in how I concatenate the final script or how I use it. I guess some confusion between ES6 `import` and CJS `require`.

pavelsavara commented:
/azp run runtime-manual

pavelsavara committed:
return back the hack for _MonoSelectRuntimeComponents

pavelsavara committed:
folders in nupkg are hard

pavelsavara commented:
`[FAIL] System.Diagnostics.Tests.StopwatchTests.GetTimestamp` is https://github.com/dotnet/runtime/issues/62021

pavelsavara commented:
/azp run runtime-manual

radical commented on src/mono/sample/wasm/browser-es6/Wasm.Browser.Sample.csproj:
```diff
@@ -0,0 +1,20 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Rename this to `Wasm.Browser.ES6.Sample.csproj`, else it will just conflict with the existing browser sample project, and only one of them would get picked up for testing.

` Zipping directory "/__w/1/s/src/mono/sample/wasm/browser-es6/bin/Release/AppBundle/" to "/__w/1/s/src/mono/sample/wasm/browser-es6/bin/browser-wasm/Wasm.Browser.Sample.zip".`
`Zipping directory "/__w/1/s/src/mono/sample/wasm/browser/bin/Release/AppBundle/" to "/__w/1/s/src/mono/sample/wasm/browser/bin/browser-wasm/Wasm.Browser.Sample.zip".`

radical commented on src/mono/sample/wasm/browser-legacy/Wasm.Browser.LegacySample.csproj:
```diff
@@ -0,0 +1,21 @@
+    <DebugSymbols>true</DebugSymbols>
+    <DebugType>embedded</DebugType>
+    <WasmDebugLevel>1</WasmDebugLevel>
+    <WasmBuildNative>true</WasmBuildNative>
```
Why does this need relinking?

pavelsavara replied:
It does not, right now. This was just easy way how to test re-link of CJS.
After ES6 is default it will need re-link.

pavelsavara committed:
feedback

pavelsavara committed:
Merge branch 'main' into wasm_modularization7

# Conflicts:
#	src/libraries/Native/Unix/System.Native/pal_random.js
#	src/mono/wasm/wasm.proj
#	src/native/libs/System.Native/pal_random.js
#	src/native/libs/System.Native/pal_random.lib.js

pavelsavara commented on src/mono/wasm/runtime/modularize-cjs/dotnet.cjs.post.js:
```diff
@@ -0,0 +1,3 @@
+createDotnetRuntime.ready = createDotnetRuntime.ready.then(() => {
+    return { MONO, BINDING, INTERNAL, Module };
```
note to self: return api object from exports.ts

pavelsavara commented on src/mono/wasm/runtime/modularize-es6/dotnet.es6.post.js:
```diff
@@ -0,0 +1,3 @@
+createDotnetRuntime.ready = createDotnetRuntime.ready.then(() => {
+    return { MONO, BINDING, INTERNAL, Module };
```
api

pavelsavara committed:
minor fixes

pavelsavara committed:
Merge branch 'main' into wasm_modularization7

# Conflicts:
#	src/mono/wasm/runtime/dotnet.lib.js
#	src/mono/wasm/runtime/rollup.config.js

pavelsavara committed:
- move dotnet.d.ts one level up
- add package.json to nuget package
- unify naming case DotNet to Dotnet
- unify emscripten type definitions
- export same instance of __dotnet_exportedAPI on all places
- write dotnet.d.ts to version control, so that we could observe when it changes

pavelsavara committed:
fix merge

pavelsavara committed:
NodeJS work in progress

pavelsavara committed:
move files

pavelsavara committed:
fix

pavelsavara commented:
There was timeout on `WebAssembly.Browser.RuntimeConfig.Test` [log](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-88722ccc3b3d489b80/WebAssembly.Browser.RuntimeConfig.Test/1/console.779ce15a.log?sv=2019-07-07&se=2021-12-15T20%3A43%3A33Z&sr=c&sp=rl&sig=rz4vdbAyu3kxSqwZzhL%2BQc%2B4OMJZNxAMz6y1SoCMLUk%3D) and `WebAssembly.Browser.HotReload.Test` [log](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-88722ccc3b3d489b80/WebAssembly.Browser.HotReload.Test/1/console.204d1804.log?sv=2019-07-07&se=2021-12-15T20%3A43%3A33Z&sr=c&sp=rl&sig=rz4vdbAyu3kxSqwZzhL%2BQc%2B4OMJZNxAMz6y1SoCMLUk%3D) on some of the previous builds.

pavelsavara committed:
fix build

pavelsavara committed:
simplify TS Sample

pavelsavara committed:
improve exports, reduce console noise, fix build

pavelsavara committed:
flat sample

pavelsavara commented:
/azp run runtime-manual

pavelsavara committed:
fix test

pavelsavara commented:
/azp run runtime-manual

pavelsavara commented:
There are many fails on `System.Text.Json.Serialization.Tests` on manual pipeline. I believe they are unrelated. All non-wasm failures are unrelated too.

radical commented:
Are these failing on main too?

pavelsavara commented:
> Are these failing on main too?
I just started one, we will see
https://dev.azure.com/dnceng/public/_build/results?buildId=1487862&view=results

pavelsavara committed:
Merge branch 'main' into wasm_modularization7

pavelsavara commented:
/azp run runtime-manual

pavelsavara commented:
>> There are many fails on `System.Text.Json.Serialization.Tests` on manual pipeline.

> Are these failing on main too?

I merged the main and run both main and this branch. Apparently this branch is causing it [Log](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-3922ee645675469483/System.Text.Json.Tests/1/console.2fe393d4.log?sv=2019-07-07&se=2021-12-18T16%3A26%3A52Z&sr=c&sp=rl&sig=b7%2BbjPq5H8JKThmxUbpyIE2yvCmEC1VFG4L%2Bmymf9h0%3D) and main is OK.
```
System.InvalidCastException : Specified cast is not valid.
[17:41:49] info:    at System.Text.Json.Serialization.JsonConverter
```
and interestingly also
```
[FAIL] System.Text.Json.Serialization.Tests.StreamTests_Async.VeryLargeJsonFileTest(payloadSize: 1, ignoreNull: True, writeIndented: True)
[17:41:23] info: Assert.Equal() Failure
[17:41:23] info:                                   ↓ (pos 87)
[17:41:23] info: Expected: ···fbbb888kkk",\n      "Name": "John Doe",\n      "Username": "joh···
[17:41:23] info: Actual:   ···fbbb888kkk",\n      "Username": "johndoe",\n      "CreatedAt": ···
```
And
```
System.InvalidOperationException : The object with reference id '1' of type 'System.Collections.Generic.Dictionary`2[System.String,System.Collections.Generic.IDictionary`2[System.String,System.Collections.Specialized.OrderedDictionary]]' cannot be assigned to the type 'System.Text.Json.Serialization.Tests.ReferenceHandlerTests+ListWithGenericCycle'.
```

I have no idea right now.  Could something in JS change affect serialization order ? How ?

kg commented:
Is it even serializing an ordered container? Most dictionaries do not have a defined ordering and in some cases the ordering will vary between runs as an anti-DoS measure.

pavelsavara commented:
I can reproduce it on my local ubuntu. 

It doesn't happen in without AOT. It doesn't happen even with AOT, when you filter for a particular test method.
So, it's side effect between tests.

pavelsavara commented:
It happens with Debug version of AOT as well. I noticed interesting difference in `dotnet.js` vs non-AOT.
There are few methods extra in AOT version, dealing with exceptions. 
I don't know what it does or why or if that's even related.

![image](https://user-images.githubusercontent.com/271576/143791742-324430df-fee5-45dd-8edf-635566b3c36c.png)

pavelsavara commented:
It seems that my hack for **linking out eventpipe is enough** to cause it the json issues.
Here I created branch just with that change https://github.com/dotnet/runtime/pull/62116 and I'm testing on CI.

pavelsavara commented:
Created https://github.com/dotnet/runtime/issues/62125 for the issue.

pavelsavara committed:
try without dealing with eventPipe issue

pavelsavara committed:
type exports for Blazor

pavelsavara committed:
keep sha local

pavelsavara committed:
types

pavelsavara commented:
/azp run runtime-manual

pavelsavara committed:
move type definitions

pavelsavara committed:
remove typescript sample

pavelsavara committed:
Merge branch 'main' into wasm_modularization7

pavelsavara commented:
/azp run runtime-manual

pavelsavara commented:
- I removed the TS console sample as it's not 100% ready yet.
- I also removed the hack-fix for eventPipe linking, perhaps Ankit was right and I don't need it fixed for this PR.
@kg @radical could you guys please give it another review ?

radical commented on src/mono/wasm/test-main.js:
```diff
@@ -244,21 +253,24 @@ function set_exit_code(exit_code, reason) {
     if (is_browser) {
-        const stack = (new Error()).stack.replace(/\n/g, "").replace(/[ ]*at/g, " at").replace(/https?:\/\/[0-9.:]*/g, "").replace("Error", "");
-        const messsage = `Exit called with ${exit_code} when isXUnitDoneCheck=${isXUnitDoneCheck} isXmlDoneCheck=${isXmlDoneCheck} WS.bufferedAmount=${consoleWebSocket.bufferedAmount} ${stack}.`;
+        // const stack = (new Error()).stack.replace(/\n/g, "").replace(/[ ]*at/g, " at").replace(/https?:\/\/[0-9.:]*/g, "").replace("Error", "");
+        // const messsage = `Exit called with ${exit_code} when isXUnitDoneCheck=${isXUnitDoneCheck} isXmlDoneCheck=${isXmlDoneCheck} WS.bufferedAmount=${consoleWebSocket.bufferedAmount} ${stack}.`;
```
can these be removed now?

pavelsavara replied:
removed

radical commented on src/mono/wasm/test-main.js:
```diff
@@ -244,21 +253,24 @@ function set_exit_code(exit_code, reason) {
 
-        console.log('WS: ' + messsage);
-        originalConsole.log('CDP: ' + messsage);
+        // console.log('WS: ' + messsage);
+        // originalConsole.log('CDP: ' + messsage);
```
and this

pavelsavara replied:
removed

radical commented on src/mono/wasm/wasm.proj:
```diff
@@ -180,16 +181,30 @@
           DestinationFolder="$(NativeBinDir)src"
           SkipUnchangedFiles="true" />
 
+    <Copy SourceFiles="runtime/cjs/dotnet.cjs.pre.js;
+                       runtime/cjs/dotnet.cjs.lib.js;
+                       runtime/cjs/dotnet.cjs.post.js;
+                       runtime/cjs/dotnet.cjs.extpost.js;"
+          DestinationFolder="$(NativeBinDir)src/cjs"
+          SkipUnchangedFiles="true" />
+
+    <Copy SourceFiles="runtime/es6/dotnet.es6.pre.js;
+                       runtime/es6/dotnet.es6.lib.js;
+                       runtime/es6/dotnet.es6.post.js;"
+          DestinationFolder="$(NativeBinDir)src/es6"
+          SkipUnchangedFiles="true" />
```
indenting

radical reviewed (changes requested):
I reviewed the parts that I could (so, skipped js, and ts!). looking good.
Just some suggestions.

pavelsavara committed:
Update src/mono/wasm/runtime/modularize-dotnet.md

Co-authored-by: Ankit Jain <radical@gmail.com>

pavelsavara committed:
Update eng/liveBuilds.targets

Co-authored-by: Ankit Jain <radical@gmail.com>

pavelsavara commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:
```diff
@@ -218,10 +219,17 @@
-    <PlatformManifestFileEntry Include="dotnet.d.ts" IsNative="true" />
-    <PlatformManifestFileEntry Include="library-dotnet.js" IsNative="true" />
-    <PlatformManifestFileEntry Include="pal_random.js" IsNative="true" />
+    <PlatformManifestFileEntry Include="package.json" IsNative="true" />
```
So that you could use it to install it as npm package into your user code. Also it enables .d.ts type definitions if you do that. 
This is just draft right now as it doesn't solve AOT use case yet.

pavelsavara commented on src/mono/sample/wasm/browser-legacy/Wasm.Browser.LegacySample.csproj:
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
`browser-legacy` is creating `globalThis.Module` object. The other one is using `createDotnetRuntime` function

pavelsavara commented on src/mono/sample/mbr/browser/main.js:
```diff
@@ -7,7 +7,7 @@ var Module = {
         MONO.config.environment_variables["DOTNET_MODIFIABLE_ASSEMBLIES"] = "debug";
     },
-    onDotNetReady: function () {
+    onDotnetReady: function () {
```
my quick survey told me Dotnet is more common, but I don't have strong opinion. 
@lewing this would stick with us going forward, what should it be ?

pavelsavara replied:
I will eventually rename it in next PR

pavelsavara commented on src/mono/wasm/runtime/.eslintrc.js:
```diff
@@ -16,7 +16,12 @@ module.exports = {
+    "ignorePatterns": [
+        "node_modules/**/*.*",
+        "bin/**/*.*",
+        "cjs/*.js",
+        "es6/*.js",
```
They are snippets not full files and so they may be not `lint` nice

pavelsavara commented on src/mono/wasm/runtime/CMakeLists.txt:
```diff
@@ -24,8 +24,8 @@ target_link_libraries(dotnet
 set_target_properties(dotnet PROPERTIES
-    LINK_DEPENDS "${NATIVE_BIN_DIR}/src/emcc-default.rsp;${NATIVE_BIN_DIR}/src/runtime.iffe.js;${SOURCE_DIR}/library-dotnet.js;${SYSTEM_NATIVE_DIR}/pal_random.js"
-    LINK_FLAGS "@${NATIVE_BIN_DIR}/src/emcc-default.rsp ${CONFIGURATION_LINK_FLAGS} -DENABLE_NETCORE=1 --extern-pre-js ${NATIVE_BIN_DIR}/src/runtime.iffe.js --js-library ${SOURCE_DIR}/library-dotnet.js --js-library ${SYSTEM_NATIVE_DIR}/pal_random.js"
+    LINK_DEPENDS "${NATIVE_BIN_DIR}/src/emcc-default.rsp;${NATIVE_BIN_DIR}/src/cjs/dotnet.cjs.pre.js;${NATIVE_BIN_DIR}/src/cjs/runtime.cjs.iffe.js;${NATIVE_BIN_DIR}/src/cjs/dotnet.cjs.lib.js;${NATIVE_BIN_DIR}/src/pal_random.lib.js;${NATIVE_BIN_DIR}/src/cjs/dotnet.cjs.post.js;${NATIVE_BIN_DIR}/src/cjs/dotnet.cjs.extpost.js;"
+    LINK_FLAGS "@${NATIVE_BIN_DIR}/src/emcc-default.rsp ${CONFIGURATION_LINK_FLAGS} -DENABLE_NETCORE=1 --extern-pre-js ${NATIVE_BIN_DIR}/src/cjs/runtime.cjs.iffe.js --pre-js ${NATIVE_BIN_DIR}/src/cjs/dotnet.cjs.pre.js --js-library ${NATIVE_BIN_DIR}/src/cjs/dotnet.cjs.lib.js --js-library ${NATIVE_BIN_DIR}/src/pal_random.lib.js --post-js ${NATIVE_BIN_DIR}/src/cjs/dotnet.cjs.post.js --extern-post-js ...
```
@radekdoulik how could I do this ? I'll do it in next PR.

pavelsavara committed:
feedback from @radical

radical reviewed (approved):
Reviewed the bits that I could, and those look good to me. great job 👍

pavelsavara commented:
/azp run runtime-manual

maraf commented on src/mono/wasm/runtime/polyfills.ts:
```diff
@@ -0,0 +1,50 @@
+export async function fetch_like(url: string): Promise<Response> {
+    try {
+        if (typeof (globalThis.fetch) === "function") {
+            return globalThis.fetch(url, { credentials: "same-origin" });
```
Is it correct to throw away original fetch settings?

pavelsavara replied:
I just moved the method from `_fetch_asset` in `startup.ts` ... but I think it's ok.
When emscripten is calling this, they are sending the same options. 
It's also the default behavior of fetch. See https://developer.mozilla.org/en-US/docs/Web/API/Request/credentials

Maybe it's there for purpose, so that it could not be used with `include` ?

maraf commented on src/mono/wasm/runtime/types/emscripten.ts:
```diff
@@ -52,8 +52,12 @@ declare interface EmscriptenModule {
+    preInit?: (() => void | Promise<void>)[];
+    preRun?: (() => void | Promise<void>)[];
+    postRun?: (() => void | Promise<void>)[];
     onRuntimeInitialized?: () => void;
```
In `exports.ts` we are assigning a `mono_wasm_on_runtime_initialized` to it, which returns `promise`. 
Is `void` correct type or should it be the `promise`?

pavelsavara replied:
`void` is correct. In order to delay `run` we have to `addRunDependency/removeRunDependency` before/after async work.
Thanks!

maraf commented on src/mono/wasm/runtime/modularize-dotnet.md:
```diff
@@ -0,0 +1,94 @@
+        // This is called after all assets are loaded , mapping to legacy `config.loaded_cb`. 
+        // It happens during emscripten `onRuntimeInitialized` after monoVm init + globalization + assemblies. 
+        // This also matches when the top level promise is resolved. 
+        // The original emscripten `Module.ready` promise is replaced this this.
```
nit: typo 

```suggestion
        // The original emscripten `Module.ready` promise is replaced with this.
```

maraf commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -161,16 +170,43 @@ function initializeImportsAndExports(
     if (!module.onRuntimeInitialized) {
         module.onRuntimeInitialized = mono_wasm_on_runtime_initialized;
+        module.ready = module.ready.then(() => {
+            return mono_wasm_runtime_is_initialized;
```
Are we really returning here a function?
Maybe I don't understand how `module.ready` is awaited..

pavelsavara replied:
`mono_wasm_runtime_is_initialized` is a promise
`mono_wasm_on_runtime_initialized` is a method which is resolving it

maraf reviewed (commented):
- `debugger-main.js` still uses `onDotNetReady` (capital N)

pavelsavara committed:
@maraf's feedback

pavelsavara committed:
tweaks

pavelsavara commented:
/azp run runtime-manual

pavelsavara committed:
fix variables

pavelsavara commented:
/azp run runtime-manual

radical commented:
Some of the wasm.build.tests [have](https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61313-merge-024fe23ef74644f1b0/Workloads-.FlagsChangeRebuildTest/1/console.f9cb881c.log?sv=2019-07-07&se=2021-12-21T17%3A22%3A29Z&sr=c&sp=rl&sig=pHPtxOBJxpUb2w6prNSSIThtm%2BIHTiS6Z9B%2FG6Hy6UU%3D):

```
[yc2d32mb.kow]          EXEC : warning : invalid item (maybe a typo?) in EXPORTED_RUNTIME_METHODS: MONO [/datadisks/disk1/work/AAF1094C/w/BA0209FA/e/yc2d32mb.kow/rebuild_flags_Release.csproj]
[yc2d32mb.kow]          EXEC : warning : invalid item (maybe a typo?) in EXPORTED_RUNTIME_METHODS: BINDING [/datadisks/disk1/work/AAF1094C/w/BA0209FA/e/yc2d32mb.kow/rebuild_flags_Release.csproj]
[yc2d32mb.kow]          EXEC : warning : invalid item (maybe a typo?) in EXPORTED_RUNTIME_METHODS: INTERNAL [/datadisks/disk1/work/AAF1094C/w/BA0209FA/e/yc2d32mb.kow/rebuild_flags_Release.csproj]
```

pavelsavara commented:
- EXPORTED_RUNTIME_METHODS - I will fix it
- I tested debugger unit tests, only well known `DebuggerTests.HarnessTests.BrowserClose` fails.
- On CI `[FAIL] System.Diagnostics.Tests.StopwatchTests.GetTimestamp` is https://github.com/dotnet/runtime/issues/62021
- On CI `Assertion failed: native function malloc called before runtime initialization` - is when emcc `-s ASSERTIONS=1`. Blazor has the same issue I think. It works anyway, when asserts are not enabled. I will change it back to run all in `onRuntimeInitialized`, but in that case `postRun` event is running in parallel with loading of assemblies and code execution. This would be consistent with previous runtime behavior.

pavelsavara committed:
fix exports, fix early malloc

pavelsavara commented:
/azp run runtime-manual
